### ISSUE TITLE
Implement a design tokens strategy for managing all variables in a centralized place

### ISSUE THREAD

**Issue by levimcg** [indiana-university/rivet-source#214]
## Summary
As we start work on Rivet 2, I'd like to look at how we can implement a Design Tokens strategy that was [popularized by Jina Anne through Salesforce's Lightning Design system](https://www.lightningdesignsystem.com/design-tokens/).

This would allow us to centralize all of the primitive values we use (e.g. .scss variables) in one place, in a format that is portable and standard (JSON) and can be translated into any other format e.g. .scss variable, less variables, imported into design tools, etc.

## Describe the solution you'd like
I'd like to start by exploring the tool/npm package Salesforce has open sourced called Theo that was built to generate design tokens.

https://github.com/salesforce-ux/theo

I really like that you can add a lot of metadata to each token beyond just the values needed for CSS, etc. Take this example from the Duet design system mentioned below.

```json
"color_warning": {
  "originalValue": "#f7b228",
  "theme": "lahitapiola",
  "name": "color_warning",
  "order": "09",
  "value": "rgb(247, 178, 40)",
  "status": "ready",
  "title": "Status Warning",
  "type": "color",
  "description": "Warning orange (non-accessible) is only used in special cases like form validation and messaging.",
  "category": "color"
}
```

You can sort of see you could also use this kind of file to also generate documentation via some static site generator (like we use for the docs site).

## Alternatives
As an alternative there are other packages that do similar things that may be worth looking into. For example this json-scss package.

https://www.npmjs.com/package/json-to-scss

## Similar features or implementations
Good examples to look at:
- [Duet Design System](https://www.duetds.com/tokens/)
- [Salesforce Lightning](https://www.lightningdesignsystem.com/design-tokens/)
- [Shopify Polaris Design Tokens](https://github.com/Shopify/polaris-tokens)

## Unknowns
One thing I'm unsure of is whether this should be part of `rivet-source` or if we should go ahead and split it off/isolate it as its own package that can be installed in any project where it's needed. For example I could see it being very useful to use on the next version of `rivet-docs` to generate color, spacing, typography documentation, etc.

I'm definitely leaning toward the separate package (similar to the way we're handling our eslint and stylelint configs).

**Comment by levimcg** [comment]
@zttodd @illusivesunrae Adding a little more detail to this issue after having some time to think on it.

After chatting with Zach about his research on [Style Dictionary](https://amzn.github.io/style-dictionary/#/) it seems like the best tooling option in terms of documentation, flexibility, etc.

## Initial tokens/properties
To get started I made a list of initial design token categories (or _properties_) as Style Dictionary refers to them. Each of these would potential equate to a JSON file with a collection of standard values that currently exist in Rivet (typically only Sass variables at the moment).

- [Font sizes](https://rivet.iu.edu/components/layout/typography/) (type scale)
- Font styles
- Colors
- Borders
- Radius
- [Spacing values](https://rivet.iu.edu/components/layout/spacing/) (margin, padding, etc.)
- Shadows
- [Widths](https://rivet.iu.edu/components/utilities/width/)

## Implementation details
Let's plan to have the master design tokens source live in the `rivet-source` repo. The source tokens are generally a standard format like JSON that can then be converted to different formats using various Node.js-based tooling. For Rivet we can keep all of the design token values in source control and generate whatever formats we need from there.

### Potential workflow
Once we have our initial set of master design tokens in place a possible workflow might look like this:

1. Add a new value to `master-tokens.json` (or whatever the name is)
2. Run a command to regenerate design tokens using tooling mentioned above and commit new/updated files to source.
3. Rivet's Sass codebase imports the generated `variables.scss` (or whatever the names end up being) file.
4. Repeat any time we add new _master_ values/properties to the master JSON file(s)

### Design tokens distribution
Once we have the tooling and process for generating tokes in place, let's plan to ship the generated tokens as part of the `rivet-core` npm package. E.g. add `tokens` to the `"files"` array in `package.json`.

For this initial implementation I would say that maybe we limit the generated formats to:
- .scss variables, and/or .scss maps for where we need to be able to loop through and create lots of CSS classes for things like colors, spacing, typescale, etc.
- .less variables (we have quite a few teams that use less)
- A plain .json object so we can use the tokens in other tooling, like static site generators to generate documentation and examples. For example [using global data .json file sin 11ty](https://www.11ty.dev/docs/data-global/).
- We can talk about what other formats make sense once we get the tooling in place.

This is all just an initial stab a how we might implement this. As you're working through this feel free to bring questions about naming conventions, process, etc. I've linked three good examples of implementation so make sure to take a look at those to get a feel for what we are trying to accomplish here. Beyond that, there are tons of good articles and examples out there you might want to reference to get some ideas.

**Comment by illusivesunrae** [comment]
Released with PR #294.
